File: 2016/07_Internet_Protocol_Version_7/part_1/main.c
/**
 * Author: happ_2h https://github.com/happ-2h
 *
 * Solution for Advent of Code 2016 Day 7.1
 */

#include <stdbool.h>
#include <stdio.h>

int main(int argc, char const *argv[]) {
  FILE* ifile = fopen("../input.txt", "r");

  if (ifile) {
    char buf[0xFF];
    int supportCtr = 0;

    while (fgets(buf, sizeof buf, ifile)) {
      bool isABBA     = false;
      bool isHypernet = false;

      // Stop check if remaining characters < 4
      for (const char* p_buf = buf; p_buf[3] != '\0'; ++p_buf) {
        if (*p_buf == '[')      isHypernet = true;  // Entered hypernet string
        else if (*p_buf == ']') isHypernet = false; // Left hypernet string

        if (p_buf[0] == p_buf[3] && p_buf[1] == p_buf[2] && p_buf[0] != p_buf[1]) {
          // If hypernet and ABBA, don't count it and move on
          if (isHypernet) {
            isABBA = false;
            break;
          }
          // Otherwise count it
          else isABBA = true;
        }
      }

      supportCtr += isABBA;
    }

    fprintf(stdout, "%d\n", supportCtr);

    fclose(ifile);
    ifile = NULL;
  }
  else {
    perror("Failed to read input.txt");
    return 1;
  }

  return 0;
}

File: 2016/07_Internet_Protocol_Version_7/part_1/main.js
/**
 * Author: happ_2h https://github.com/happ-2h
 *
 * Solution for Advent of Code 2016 Day 7.1
 *
 * FIXME
 *   More counted than it should
 */

let input = null;

try {
  input = require("fs").readFileSync("../input.txt", "utf-8").split('\n');
} catch (err) {
  console.error(err);
}

if (input !== null) {
  const isABBA = (str) => {
    // Check if has reverse pair
    const regRes = str.match(/([a-z])([a-z])\2\1/g);

    if (regRes) {
      // Check if different characters
      if (regRes[0][0] != regRes[0][1]) {
        return true;
      }
    }
    // No ABBA pattern
    return false;
  };

  let supportCtr = 0;

  for (const str of input) {
    const splitInput = str.split(/\[(.*?)\]/);
    let strOut = ""; // Ouside brackets
    let strIn = "";  // Inside brackets

    // Concatenate strings
    for (let i = 0; i < splitInput.length; ++i) {
      // Even: outside brackets
      if ((i&1) == 0) {
        strOut += splitInput[i];
      }
      // Odd: inside brackets
      else {
        strIn += splitInput[i];
      }
    }

    // Validate IPv7
    if (!isABBA(strIn) && isABBA(strOut)) ++supportCtr;
  }

  console.log(supportCtr);
}
else console.error("Failed to read input.txt");